### PR TITLE
Add Dirichlet distribution

### PR DESCRIPTION
**Pull request by YoshikawaMasashi**
split from #4678.

### CODE REVIEW

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

File: chainer/distributions/dirichlet.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        return sum_mod.sum(self.alpha, axis=-1)
+
+    @property
+    def batch_shape(self):
+        return self.alpha.shape[:-1]
+
+    @property
+    def entropy(self):
+        return sum_mod.sum(lgamma.lgamma(self.alpha), axis=-1) \
+            - lgamma.lgamma(self.alpha0) \
```
Comment (toslunar) [inline chainer/distributions/dirichlet.py]: The multivariate log beta function (not necessarily a `FunctionNode`) could be useful in `entropy`, `log_prob`, and kl.

File: chainer/distributions/dirichlet.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    def __init__(self, alpha):
+        self.__alpha = chainer.as_variable(alpha)
+        self.__k = alpha.shape[-1]
+
+    @property
+    def alpha(self):
+        return self.__alpha
+
+    @property
+    def k(self):
```
Comment (toslunar) [inline chainer/distributions/dirichlet.py]: I'd like to encourage users to access this via `event_shape`, instead of providing the property `k`.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Pull request merged**